lib/App/index.js

```diff
@@ -116,6 +117,38 @@ class App {
       throw new Error('Invalid compatibility');
     }
 
+    if (appJson.esm === true && semver.lt(semver.coerce(appJson.compatibility), '12.0.0')) {
+      throw new Error('ESM apps require a compatibility of at least >=12.0.0');
+    }
+
+    const checkMjs = async (path) => {
+      await fs.promises.access(path).then(() => {
+        if (semver.lt(semver.coerce(appJson.compatibility), '12.0.0')) {
+          throw new Error(`ESM apps require a compatibility of at least >=12.0.0. (${path})`);
+        }
+      }).catch(err => {
+        if (err.code !== 'ENOENT') {
+          throw err;
+        }
+      });
+    };
+
+    const appFilePath = join(this._path, 'app.mjs');
+    const apiFilePath = join(this._path, 'api.mjs');
+
+    await checkMjs(appFilePath);
+    await checkMjs(apiFilePath);
+
+    if (Array.isArray(appJson.drivers)) {
+      for (const driver of appJson.drivers) {
+        const driverFilePath = join(this._path, 'drivers', driver.id, 'driver.mjs');
+        const deviceFilePath = join(this._path, 'drivers', driver.id, 'device.mjs');
+
+        await checkMjs(driverFilePath);
+        await checkMjs(deviceFilePath);
+      }
+    }
```

Comment:
This still leaves the possibility for other (unknown) files being .mjs while compatiblity is not set to >=12. Not sure how/if we could eliminate that.

Reply:
How would they even import those files?